### PR TITLE
Add RCE and SSTI Active Scan Scripts

### PR DESCRIPTION
**Pull request by knassar702**
hi team

i made remote code execution& Template injection scripts for Active scan script 

## RCE 

bypass
```
spaces,',`,/,\,<,>,?,&,|
```

my web app for testing
```python
#!/usr/bin/env python3
from flask import Flask,request
from os import popen
app = Flask(__name__)
@app.route('/')
def index():
    r = request.args.get('name',None)
    if r:
        r = r.replace('&','').replace('|','').replace('`','').replace(';','').replace(' ','')
        try:
            R = popen(f'ping -c 1 {r}').read().rstrip()
            print(R)
            if len(R) < 2:
                r = f'{r} | Error ..!'
                return r
            return R
        except Exception as e:
            return r
    return '''
What is your name .?
<hr>
<form action="/" method="GET">
<input type="text" name="name">
<input type="submit" value="Go">
</form>
    '''
if __name__ == '__main__':
    app.run()

```
Screenshot at 2020-09-02 03-37-45

## SSTI 
91222171-4fa51a00-e70e-11ea-8984-eae807c2c0b0

best regards :heart: 
```

### CODE REVIEW

**Comment by kingthorin** [comment]
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/develop/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

File: active/RCE.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+"""
+follow_redirects = False
+def scanNode(sas, msg):
+  # Copy requests before reusing them
+  msg = msg.cloneRequest();
+
+  # sendAndReceive(msg, followRedirect, handleAntiCSRFtoken)
+  sas.sendAndReceive(msg, follow_redirects, False);
+
+  # Test the responses and raise alerts as below
```
Comment (kingthorin) [inline active/RCE.py:16]:
```suggestion
  pass
```

File: active/RCE.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+rce_payloads = {
+    '''
+cat${IFS}/etc/passwd''':'root:x:0:0',
+    '''
+cat /etc/passwd''':'root:x:0:0',
+    '''
+uname''':'gid='
+    }
```
Comment (kingthorin) [inline active/RCE.py:25]: This may as well be moved up with `follow_redirects`

File: active/RCE.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+def scan(sas, msg, param, value):
+  # Debugging can be done using print like this
+#  print('scan called for url=' + msg.getRequestHeader().getURI().toString() +
+#    ' param=' + param + ' value=' + value);
```
Comment (kingthorin) [inline active/RCE.py:30]:
```suggestion
```

File: active/RCE.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  for P,message in rce_payloads.items():
```
Comment (kingthorin) [inline active/RCE.py:31]:
```suggestion
  for payload,message in rce_payloads.items():
```

File: active/RCE.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  # Copy requests before reusing them
+    msg = msg.cloneRequest();
+  # setParam (message, parameterName, newValue)
+    sas.setParam(msg, param, P);
```
Comment (kingthorin) [inline active/RCE.py:35]:
```suggestion
    sas.setParam(msg, param, payload);
```

File: active/RCE.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  # sendAndReceive(msg, followRedirect, handleAntiCSRFtoken)
+    sas.sendAndReceive(msg, follow_redirects, False);
+  # Test the response here, and make other requests as required
+#  print('scan called for url=' + msg.getRequestHeader().getURI().toString());
```
Comment (kingthorin) [inline active/RCE.py:39]:
```suggestion
```

File: active/SSTI.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+* Email : knassar702@gmail.com
+"""
+def scanNode(sas, msg):
+  pass
+
+ssti_payloads = {
+    'abcd{{6*6}}123':'abcd36123',
+    'abcd${6*6}123':'abcd36123',
+    'abcd<% 6*6 %>123':'abcd36123'
+    }
```
Comment (kingthorin) [inline active/SSTI.py:14]:
```suggestion

ssti_payloads = {
    'abcd{{6*6}}123':'abcd36123',
    'abcd${6*6}123':'abcd36123',
    'abcd<% 6*6 %>123':'abcd36123'
    }

def scanNode(sas, msg):
  pass
```

File: active/SSTI.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+def scan(sas, msg, param, value):
+  for payload,evidence in ssti_payloads.items():
+  # Copy requests before reusing them
+    msg = msg.cloneRequest();
+  # setParam (message, parameterName, newValue)
+    sas.setParam(msg, param, payload);
+  # sendAndReceive(msg, followRedirect, handleAntiCSRFtoken)
+    sas.sendAndReceive(msg, False, False);
+  # Test the response here, and make other requests as required
```
Comment (kingthorin) [inline active/SSTI.py:24]:
```suggestion
```

File: active/RCE.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    if message in msg.getResponseBody().toString():
+  	# Change to a test which detects the vulnerability
```
Comment (kingthorin) [inline active/RCE.py:41]:
```suggestion
```

**Comment by knassar702** [comment]
@kingthorin 
Done :)

**Review by kingthorin — APPROVED** [review]
LGTM

**Comment by kingthorin** [comment]
You can just drop the full encode script, The Encoder add-on now bundles that functionality, per: https://github.com/zaproxy/zap-extensions/pull/2538 since releases 0.3.0: https://github.com/zaproxy/zap-extensions/releases/tag/encoder-v0.3.0

**Comment by kingthorin** [comment]
@knassar702 do you intend to tackle the outstanding items?

**Comment by kingthorin** [comment]
`Encode/full_urlencoding.py` is the script it has an issue with, which should have been removed per previous comments.

If for some reason you're really convinced this is still needed I think it'll need some slight modification, like:
- https://stackoverflow.com/questions/66567325/how-to-access-encodedecoderesult-from-jython-zap
- https://github.com/zaproxy/zap-extensions/pull/2755

**Comment by knassar702** [comment]
hi @kingthorin 
I added the file by mistake (` git add .`),  I will delete it now :)

**Comment by kingthorin** [comment]
I've squashed all the commits, rebased current, and updated the CHANGELOG. Before you do anything further please:
```console
git fetch origin
git reset --hard origin/master
```
(The above assumes origin is https://github.com/knassar702/community-scripts.git)

In the future we'd suggest creating a new branch before making any changes. That way you always have a good default branch to go back to, instead of having your primary/default branch busy with changes.

**Review by kingthorin — APPROVED** [review]

**Comment by kingthorin** [comment]
If people are otherwise happy with this PR I’m happy to fix the conflict.

**Comment by kingthorin** [comment]
@psiinon @thc202 @ricekot ^ ?

**Comment by psiinon** [comment]
Sounds good to me. Can you also fix the formatting? :) The comments need to be indented - I think it makes the code much harder to read when they are the same level as the surrounding conditionals / loops ;)

**Comment by kingthorin** [comment]
Sure

**Comment by kingthorin** [comment]
Done, I hope.

File: active/RCE.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    sas.setParam(msg, param, payload);
+  # sendAndReceive(msg, followRedirect, handleAntiCSRFtoken)
+    sas.sendAndReceive(msg, follow_redirects, False);
+  # Test the response here, and make other requests as required
+    if message in msg.getResponseBody().toString():
+    # raiseAlert(risk, int reliability, String name, String description, String uri,
+    #		String param, String attack, String otherInfo, String solution, String evidence,
+    #		int cweId, int wascId, HttpMessage msg)
+    # risk: 0: info, 1: low, 2: medium, 3: high
+    # reliability: 0: falsePassitive, 1: suspicious, 2: warning
```
Comment (kingthorin) [inline active/RCE.py:35]:
```suggestion
    # reliability: 0: falsePositive, 1: suspicious, 2: warning
```

File: active/RCE.py [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+def scan(sas, msg, param, value):
+  for payload,message in rce_payloads.items():
+  # Copy requests before reusing them
+    msg = msg.cloneRequest();
+  # setParam (message, parameterName, newValue)
+    sas.setParam(msg, param, payload);
+  # sendAndReceive(msg, followRedirect, handleAntiCSRFtoken)
+    sas.sendAndReceive(msg, follow_redirects, False);
+  # Test the response here, and make other requests as required
+    if message in msg.getResponseBody().toString():
```
Comment (kingthorin) [inline active/RCE.py:30]: Comment indentation should match content.

File: active/RCE.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    # raiseAlert(risk, int reliability, String name, String description, String uri,
+    #		String param, String attack, String otherInfo, String solution, String evidence,
+    #		int cweId, int wascId, HttpMessage msg)
+    # risk: 0: info, 1: low, 2: medium, 3: high
+    # reliability: 0: falsePassitive, 1: suspicious, 2: warning
+        sas.raiseAlert(3, 3, ' Remote Code Execution (Script)', 'Attack technique used for unauthorized execution of operating system commands. This attack is possible when an application accepts untrusted input to build operating system commands in an insecure manner involving improper data sanitization, and/or improper calling of external programs.',msg.getRequestHeader().getURI().toString(), param, P, '', '',message, 0, 0, msg);
```
Comment (kingthorin) [inline active/RCE.py:36]: `P` was meant to be payload? For the alert's attack value?
This should probably use an alert builder for better clarity.

**Review by kingthorin — COMMENTED** [review]
I'll tackle this later this evening.

**Comment by kingthorin** [comment]
I'm not sure what to do about the LGTM thing. It seems to be specific to python 2. Our tests pass with the code formatted as-is. 

I "guess" LGTM would be happy with it all on one line, but it's horrid to read like that....

**Comment by ricekot** [comment]
Not sure, but you could try putting the whole thing in parantheses `()` and remove the backslashes?

**Comment by kingthorin** [comment]
Fully bracketed didn't work either.

**Comment by kingthorin** [comment]
We've discussed and done some testing on the LGTM finding via IRC and have agreed to ignore it.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**